extensions/amp-viewer-integration/0.1/highlight-handler.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Messaging} from './messaging/messaging';
+import {Services} from '../../../src/services';
+import {findSentences, markTextRangeList} from './findtext';
+import {parseJson} from '../../../src/json';
+import {parseQueryString} from '../../../src/url';
+
+/**
+ * The message name sent by viewers to dismiss highlights.
+ * @type {string}
+ */
+const HIGHLIGHT_DISMISS = 'highlightDismiss';
+
+/**
+ * The length limit of highlight param to avoid parsing
+ * a incredibley large string as JSON. The limit is 100kB.
+ * @type {number}
+ */
+const HIGHLIGHT_PARAM_LENGTH_LIMIT = 100 << 10;
+
+/**
+ * The limit of # of sentences to highlight.
+ * @type {number}
+ */
+const NUM_SENTENCES_LIMIT = 15;
+
+/**
+ * The length limit of one sentence to highlight.
+ * @type {number}
+ */
+const NUM_ALL_CHARS_LIMIT = 1500;
+
+/**
+ * Returns highlight param in the URL hash.
+ * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+ * @return {?JsonObject}
+ */
+export const getHighlightParam = function(ampdoc) {
+  const param = parseQueryString(ampdoc.win.location.hash)['highlight'];
+  if (!param || param.length > HIGHLIGHT_PARAM_LENGTH_LIMIT) {
+    return null;
+  }
+  const highlight = parseJson(param);
+  const sens = highlight['s'];
+  if (sens) {
+    if (sens.length > NUM_SENTENCES_LIMIT) {
+      // Too many sentences, do nothing for safety.
+      return null;
+    }
+    let sum = 0;
+    for (let i = 0; i < sens.length; i++) {
+      const sen = sens[i];
+      if (!sen) {
+        continue;
+      }
+      sum += sen.length;
+      if (sum > NUM_ALL_CHARS_LIMIT) {
+        // Too many chars, do nothing for safety.
+        return null;
+      }
+    }
+  }
+  return highlight;
+};
+
+/**
+ * HighlightHandler reads highlight parameter from URL and
+ * highlights specified text in AMP documents.
+ */
+export class HighlightHandler {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!JsonObject} highlightInfo The highlighting info in JSON.
```

Comment:
Consider passing only an array of strings here. `JsonObject` is opaque to structure, whereas `!Array<string>` is clearer.

Reply:
If so, would you mind mapping the `JsonObject` to a structured typedef when parsing initially? That way the structure of the JSON object is explicit rather than implicit.

Reply:
Done